File: SC/Assignment1.cpp
#include <iostream>   // for input/output
#include <vector>     // to use vectors (dynamic arrays)
#include <algorithm>  // for max() and min() functions
using namespace std;


// Rule: take the maximum membership value for each element
vector<double> fuzzyUnion(const vector<double>& A, const vector<double>& B) {
    vector<double> result;  // store the union result
    for (size_t i = 0; i < A.size(); i++) {        // loop through each element
        result.push_back(max(A[i], B[i]));        
    }
    return result;  
}


// Rule: take the minimum membership value for each element
vector<double> fuzzyIntersection(const vector<double>& A, const vector<double>& B) {
    vector<double> result;  // store the intersection result
    for (size_t i = 0; i < A.size(); i++) {        // loop through each element
        result.push_back(min(A[i], B[i]));         
    }
    return result;  // return intersection result
}


// Rule: subtract membership value from 1
vector<double> fuzzyComplement(const vector<double>& A) {
    vector<double> result;  
    for (double val : A) {                  // loop through each membership value
        result.push_back(1.0 - val);        // µA' = 1 - µA
    }
    return result;  // return complement result
}



void printSet(const vector<double>& set) {
    cout << "{ ";
    for (double val : set) {      // loop through each membership value
        cout << val << " ";       
    }
    cout << "}" << endl;         
}

int main() {
    // Example fuzzy sets A and B with membership values
    // Each value is between 0 and 1
    vector<double> A = {0.2, 0.7, 0.5, 1.0};  
    vector<double> B = {0.6, 0.3, 0.8, 0.4};  

    // Print original sets
    cout << "Set A: "; printSet(A);
    cout << "Set B: "; printSet(B);

    // Perform fuzzy operations
    vector<double> unionSet = fuzzyUnion(A, B);              // A ∪ B
    vector<double> intersectionSet = fuzzyIntersection(A, B); // A ∩ B
    vector<double> complementA = fuzzyComplement(A);          // A'

    // Print results
    cout << "Union (A ∪ B): "; printSet(unionSet);
    cout << "Intersection (A ∩ B): "; printSet(intersectionSet);
    cout << "Complement (A'): "; printSet(complementA);

    return 0;  // program ends successfully
}

File: SC/Assignment1.java
import java.util.ArrayList;
import java.util.Scanner;

public class Assignment1 {

    // Union: max(A[i], B[i])
    public static ArrayList<Double> fuzzyUnion(ArrayList<Double> A, ArrayList<Double> B) {
        ArrayList<Double> result = new ArrayList<>();
        for (int i = 0; i < A.size(); i++) {
            result.add(Math.max(A.get(i), B.get(i)));
        }
        return result;
    }

    // Intersection: min(A[i], B[i])
    public static ArrayList<Double> fuzzyIntersection(ArrayList<Double> A, ArrayList<Double> B) {
        ArrayList<Double> result = new ArrayList<>();
        for (int i = 0; i < A.size(); i++) {
            result.add(Math.min(A.get(i), B.get(i)));
        }
        return result;
    }

    // Complement: 1 - A[i]
    public static ArrayList<Double> fuzzyComplement(ArrayList<Double> A) {
        ArrayList<Double> result = new ArrayList<>();
        for (double val : A) {
            result.add(1.0 - val);
        }
        return result;
    }

    public static void printSet(ArrayList<Double> set) {
        System.out.print("{ ");
        for (double val : set) {
            System.out.print(val + " ");
        }
        System.out.println("}");
    }

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of elements in fuzzy sets A and B: ");
        int n = sc.nextInt();

        ArrayList<Double> A = new ArrayList<>();
        ArrayList<Double> B = new ArrayList<>();

        // Input for A
        System.out.println("Enter values of Set A:");
        for (int i = 0; i < n; i++) {
            A.add(sc.nextDouble());
        }

        // Input for B
        System.out.println("Enter values of Set B:");
        for (int i = 0; i < n; i++) {
            B.add(sc.nextDouble());
        }

        System.out.print("Set A: ");
        printSet(A);

        System.out.print("Set B: ");
        printSet(B);

        ArrayList<Double> unionSet = fuzzyUnion(A, B);
        ArrayList<Double> intersectionSet = fuzzyIntersection(A, B);
        ArrayList<Double> complementA = fuzzyComplement(A);

        System.out.print("Union (A ∪ B): ");
        printSet(unionSet);

        System.out.print("Intersection (A ∩ B): ");
        printSet(intersectionSet);

        System.out.print("Complement (A'): ");
        printSet(complementA);

        sc.close();
    }
}
